src/NoSelfTransformer.ts

```diff
@@ -0,0 +1,12 @@
+import * as ts from "typescript";
+
+export function noSelfTransformerFactory<T extends ts.Node>(_context: ts.TransformationContext): ts.Transformer<T> {
```

Comment:
- Making transformer handle any kind of node defines looser constraint than it could and moves compile-time check to runtime
- You can use function expressions to specify full signature type
- This transformer doesn't handle bundles
- Even though it's a part of a type, IMO `Factory` is unnecessary there, such name better fits a function with `(program: ts.Program) => ts.TransformerFactory` signature (even though it actually is `FactoryFactory`).

```ts
const transformSourceFile: ts.Transformer<ts.SourceFile> = node => {
  const empty = ts.createEmptyStatement();
  ts.addSyntheticLeadingComment(empty, ts.SyntaxKind.MultiLineCommentTrivia, "* @noSelfInFile ", true);
  return ts.updateSourceFileNode(node, [empty, ...node.statements], node.isDeclarationFile);
};

export const noSelfTransformer: ts.TransformerFactory<ts.SourceFile | ts.Bundle> = () => node =>
    ts.isBundle(node) ? ts.updateBundle(node, node.sourceFiles.map(transformSourceFile)) : transformSourceFile(node);
```

For bundles it generates declarations like

```ts
 /** @noSelfInFile */
declare module "foo" { }
/** @noSelfInFile */
declare module "bar" { }
```

Not sure how it works now,

Reply:
Thanks a bunch for these notes and code